File: plugins/pushover-commander/skills/_lib/batch_create_pushover_apps.py
#!/usr/bin/env python3
"""Batch-create new Pushover apps from createnew.json: create-app (name+desc) then edit-app (icon).
Run from the _lib dir so `pushover_headless_web_control` imports. Records new tokens (last6) for the inventory."""
import json
import os

from playwright.sync_api import sync_playwright, Error as PWError
from pushover_headless_web_control import login, create_app, edit_app

EMAIL = os.environ["PO_EMAIL"]; PW = os.environ["PO_PW"]
PLAN_PATH = os.environ.get("CREATE_PLAN", "/tmp/po_sounds/createnew.json")
with open(PLAN_PATH, encoding="utf-8") as f:
    plan = json.load(f)

results = []
with sync_playwright() as p:
    b = p.chromium.launch(channel="chrome", headless=True)
    pg = b.new_context(viewport={"width": 1100, "height": 1700}).new_page()
    pg.on("dialog", lambda d: d.accept())
    login(pg, EMAIL, PW)
    for item in plan:
        try:
            r = create_app(pg, item["new_name"], item["desc"], "", True)  # reveal=True -> full token
            slug = r.get("app_url", "").rsplit("/", 1)[-1]
            token = r.get("token") or ""
            icon = item.get("icon")
            iconed = False
            if slug and icon and os.path.exists(icon):
                edit_app(pg, slug, None, None, None, icon)
                iconed = True
            results.append({"name": item["new_name"], "created": r.get("created"),
                            "slug": slug, "token_last6": token[-6:] if token else None, "icon": iconed})
        except (PWError, OSError) as e:
            results.append({"name": item["new_name"], "error": str(e)[:140]})

print(json.dumps(results, indent=2, ensure_ascii=False))
print(f"\ncreated {sum(1 for r in results if r.get('created'))}/{len(plan)}")

File: plugins/pushover-commander/skills/_lib/batch_create_pushover_apps.ts
#!/usr/bin/env bun
/**
 * batch_create_pushover_apps.ts — batch-create Pushover apps from a plan JSON:
 * create-app (name + description) then optionally edit-app (upload icon).
 * Records each new app's last-6 token chars for the inventory.
 *
 * Plan path: $CREATE_PLAN (default /tmp/po_sounds/createnew.json), an array of
 *   { "new_name": string, "desc": string, "icon"?: string }.
 *
 * Ported from batch_create_pushover_apps.py — reuses the exported session +
 * create/edit helpers from pushover_headless_web_control.ts (single source).
 */

import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import process from "node:process";
import {
  blankOptions,
  createApp,
  editApp,
  login,
  resolveCredentials,
  withDashboard,
} from "./pushover_headless_web_control.ts";

interface PlanItem {
  readonly new_name: string;
  readonly desc: string;
  readonly icon?: string;
}

interface BatchResult {
  readonly name: string;
  readonly created?: boolean;
  readonly slug?: string;
  readonly token_last6?: string | null;
  readonly icon?: boolean;
  readonly error?: string;
}

const lastSegment = (value: string): string => value.split("/").at(-1) ?? "";

async function createOne(
  pg: Parameters<Parameters<typeof withDashboard>[1]>[0],
  item: PlanItem,
  userKey: string,
): Promise<BatchResult> {
  try {
    const result = await createApp(
      pg,
      blankOptions({ name: item.new_name, desc: item.desc, reveal: true }),
      userKey,
    );
    const slug = lastSegment(String(result.app_url ?? ""));
    const token = typeof result.token === "string" ? result.token : "";
    let iconed = false;
    if (slug && item.icon && existsSync(item.icon)) {
      await editApp(pg, blankOptions({ slug, icon: item.icon }));
      iconed = true;
    }
    return {
      name: item.new_name,
      created: result.created === true,
      slug,
      token_last6: token ? token.slice(-6) : null,
      icon: iconed,
    };
  } catch (error) {
    return { name: item.new_name, error: (error instanceof Error ? error.message : String(error)).slice(0, 140) };
  }
}

async function main(): Promise<void> {
  const creds = resolveCredentials();
  const planPath = process.env.CREATE_PLAN ?? "/tmp/po_sounds/createnew.json";
  const plan = JSON.parse(await readFile(planPath, "utf8")) as PlanItem[];

  const results = await withDashboard(true, async (pg) => {
    await login(pg, creds);
    const collected: BatchResult[] = [];
    for (const item of plan) {
      collected.push(await createOne(pg, item, creds.userKey));
    }
    return collected;
  });

  process.stdout.write(`${JSON.stringify(results, null, 2)}\n`);
  const created = results.filter((r) => r.created).length;
  process.stdout.write(`\ncreated ${created}/${plan.length}\n`);
}

main().catch((error: unknown) => {
  process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
  process.exit(1);
});
